app/services/open_humans_service.rb

```diff
@@ -0,0 +1,111 @@
+# frozen_string_literal: true
+
+class OpenHumansService
+  # used to do the interfacing w/ Open Humans
+
+  def get_access_tokens(user, code)
+    # authenticate w/ our client id/secret against API
+    # post with the key the user provided us with.
+    url = URI.parse('https://www.openhumans.org/oauth2/token/')
+    req = Net::HTTP::Post.new(url.request_uri)
+    req.basic_auth ENV.fetch('OH_client_id'), ENV.fetch('OH_client_secret')
+    req.set_form_data('grant_type' => 'authorization_code',
+                      'code' => code,
+                      'redirect_uri' => 'http://localhost:3000/openhumans/authorize')
+    # set up request to use https
+    http = Net::HTTP.new(url.host, url.port)
+    http.use_ssl = (url.scheme == 'https')
+    # do actual request, get the json it returns
+    response = http.request(req)
+    response_json = JSON.parse(response.body)
+    update_tokens(user.open_humans_profile, response_json)
+  end
+
+  def refresh_token(user)
+    oh_profile = user.open_humans_profile
+    url = URI.parse('https://www.openhumans.org/oauth2/token/')
+    req = Net::HTTP::Post.new(url.request_uri)
+    req.set_form_data('grant_type' => 'refresh_token', 'refresh_token' => oh_profile.refresh_token)
+    http = Net::HTTP.new(url.host, url.port)
+    http.use_ssl = (url.scheme == 'https')
+    response = http.request(req)
+    response_json = JSON.parse(response.body)
+    update_tokens(oh_profile, response_json)
+  end
+
+  def set_open_humans_ids(oh_profile)
+    uri = URI.parse('https://www.openhumans.org/api/direct-sharing/project/exchange-member/')
+    uri_params = { access_token: oh_profile.access_token }
+    uri.query = URI.encode_www_form(uri_params)
+    res = Net::HTTP.get_response(uri)
+    res_json = JSON.parse(res.body)
+    oh_profile.project_member_id = res_json['project_member_id']
+    oh_profile.open_humans_user_id = res_json['username']
+    oh_profile.save
+  end
+
+  def upload_opensnp_id(user)
+    base_url = 'https://www.openhumans.org/api/direct-sharing/project/files/upload/?access_token='
+    uri = URI.parse(base_url + user.open_humans_profile.access_token)
+    boundary = '0P3NSNPH34RT50PENHUM4N5'
```

Comment:
❤️

Reply:
You could put that into a constant, too.

Reply:
done :)